patroni/ha.py

```diff
@@ -447,27 +447,33 @@ def process_sync_replication(self):
         promoting standbys that were guaranteed to be replicating synchronously.
         """
         if self.is_synchronous_mode():
-            current = self.cluster.sync.leader and self.cluster.sync.sync_standby
-            picked, allow_promote = self.state_handler.pick_synchronous_standby(self.cluster)
-            if picked != current:
-                # We need to revoke privilege from current before replacing it in the config
-                if current:
-                    logger.info("Removing synchronous privilege from %s", current)
-                    if not self.dcs.write_sync_state(self.state_handler.name, None, index=self.cluster.sync.index):
+            current = self.cluster.sync.leader and self.cluster.sync.sync_standby or []
+            picked, allow_promote = self.state_handler.pick_synchronous_standby(self.cluster, sync_node_count)
+            if set(picked) != set(current):
+                if picked:
+                    logger.info("Updating synchronous privilege from {} to {}".format(','.join(current),
+                                                                                      ','.join(picked)))
+                    if not self.dcs.write_sync_state(self.state_handler.name, picked, index=self.cluster.sync.index):
```

Comment:
This is where the dragon sits...

Let's imagine that the `current` is `(n1, n2)` and the `picked` is `(n3, n4)`.
By publishing `(n3, n4)` to DCS we are giving them a green light for the leader race.
BUT! In the `synchronous_standby_names` they are not yet declared as synchronous, therefore we can't guaranty no data loss during failover.

Order matters a lot here:
1. First you should "remove" from DCS `current` names that are not in `picked`.
2. Update `synchronous_standby_names` by putting `picked` names there.
3. Wait for some of `picked` nodes to become really synchronous.
4. Publish into DCS `picked` nodes that are already synchronous.

Basically, the [quorum state resolver](https://github.com/zalando/patroni/pull/672/files#diff-a839e95c7271edf3b62934f9e0e61f16R18) is the attempt to implement such an algorithm.

Reply:
Agree. Will fix the code.